lib/endpoint-proxy.js

```diff
@@ -521,11 +768,92 @@ class EndpointProxy {
     }
   }
 }
+class ServiceCaller {
+  constructor(client, serviceDeclaration, endpointDeclarations) {
+    if (client === void 0 || client === null) {
+      throw new Error('no database client provided');
+    }
+    if (serviceDeclaration === void 0 || serviceDeclaration === null) {
+      throw new Error('no service declaration provided');
+    } else if (!(typeof serviceDeclaration.endpointExtension === 'string' || (serviceDeclaration.endpointExtension instanceof String))) {
+      throw new Error('serviceDeclaration.endpointExtension must be a String');
+    }
+
+    if(endpointDeclarations === null || (Array.isArray(endpointDeclarations) && endpointDeclarations.length > 0)) {
+      throw new Error('endpointDeclarations cannot be null or empty.');
+    }
+
+    const db = client.getConnectionParams().database;
+    if (db !== void 0 && db !== null) {
+      throw new Error('connection for client cannot specify database');
+    }
+
+    this.$mlProxy = endpointDeclarations.reduce((proxy,  endpointDeclarations) =>
+    {
+      expandFunctionDeclaration(endpointDeclarations);
+      return proxy.withFunction(endpointDeclarations, serviceDeclaration.endpointExtension);
+    }, client.createProxy(client, serviceDeclaration)
+  );
+  }
+
+  call(functionName, args) {
+    return this.$mlProxy.execute(functionName, args, 0);
+  }
+}
+
+class EndpointCaller {
+  constructor(client, endpointDeclaration) {
+    if (client === void 0 || client === null) {
+      throw new Error('no database client provided');
+    }
+    if (endpointDeclaration === void 0 || endpointDeclaration === null) {
+      throw new Error('endpointDeclaration cannot be null.');
+    }
+    const endpoint = endpointDeclaration.endpoint;
+
+    if (endpoint === void 0 || endpoint === null) {
+      throw new Error('endpointDeclaration endpoint cannot be null.');
+    }
+    const regex = /^(.*[/])([^/]+)(.[^/.]+)$/;
+    const arrayValues = endpoint.match(regex);
+
+    const endpointDirectory = arrayValues[1];
+    const endpointFunctionName = arrayValues[2];
+    const endpointExtension = arrayValues[3];
+
+    if(endpointDirectory == null || endpointExtension == null) {
```

Comment:
Would it be defensive to check that each of endpointDirectory, endpointFunctionName, and endpointExtension is a string primitive or object (similar to the test of the endpoint property above)?